File: yunva17/Week_36/BOJ_7511_소셜네트워킹어플리케이션/BOJ_7511_소셜네트워킹어플리케이션.js
const { default: test } = require("node:test");

let input = require("fs")
  .readFileSync(
    "c:/Algorithm_Study/yunva17/Week_36/BOJ_7511_소셜네트워킹어플리케이션/input.txt"
  )
  // .readFileSync('/dev/stdin')
  .toString()
  .trim()
  .split("\n");

const test_case = Number(input[0]);
let line = 1;
let answers = [];
let parents = [];

const makeSet = (num) => {
  parents = Array.from({ length: num }, (_, i) => i);
};

const find = (a) => {
  if (parents[a] === a) return a;

  return (parents[a] = find(parents[a]));
};

const union = (a, b) => {
  const rootA = find(a);
  const rootB = find(b);

  if (rootA !== rootB) {
    parents[rootB] = rootA;
  }
};

for (let t = 0; t < test_case; t++) {
  answers.push(`Scenario ${t + 1}:`);

  const n = Number(input[line++]);
  const k = Number(input[line++]);

  makeSet(n);

  for (let i = 0; i < k; i++) {
    const [a, b] = input[line++].split(" ").map(Number);
    union(a, b);
  }

  const m = Number(input[line++]);

  for (let i = 0; i < m; i++) {
    const [a, b] = input[line++].split(" ").map(Number);
    if (find(a) === find(b)) {
      answers.push("1");
    } else {
      answers.push("0");
    }
  }

  if (t < test_case - 1) {
    answers.push("");
  }
}

console.log(answers.join("\n"));

File: l-umi/Week_36/BOJ_7511_소셜_네트워킹_어플리케이션.py
import sys
input = sys.stdin.readline

class UnionFind:
    def __init__(self, size):
        self.parent = list(range(size))
    
    def find(self, num):
        if num == self.parent[num]:
            return num  # root
        self.parent[num] = self.find(self.parent[num])
        return self.parent[num]
    
    def union(self, a, b):
        a, b = self.find(a), self.find(b)
        self.parent[b] = a

T = int(input())

for tc in range(1, T + 1):
    print(f"Scenario {tc}:")
    n = int(input()  
    k = int(input())
    uf = UnionFind(n)
    
    for _ in range(k):
        a, b = map(int, input().split()) 
        uf.union(a, b)
    
    m = int(input())  
    for _ in range(m):
        u, v = map(int, input().split())
        print("1" if uf.find(u) == uf.find(v) else "0")
    
    print()